src/QirRuntime/lib/QIR/arrays.cpp

```diff
@@ -261,10 +290,18 @@ extern "C"
         return array->dimensionSizes[dim];
     }
 
-    QirArray* quantum__rt__array_copy(QirArray* array)
+    QirArray* quantum__rt__array_copy(QirArray* array, bool force)
     {
-        assert(array != nullptr);
-        return new QirArray(array);
+        if (array == nullptr)
+        {
+            return nullptr;
+        }
+        if (force || array->userCount > 0)

```

Comment:
It would be good to have a brief comment here explaining why a new instance of `QirArray` is returned only when `force` is `true` or when `userCount` is greater than zero.

Reply:
This particular aspect isn't driven by the QirRuntime implementation but is explicitly required by the QIR spec. I'm duplicating the spec's verbatim in the quantum_rt.hpp file. I'll add a comment in the qirTypes.hpp for the corresponding members/methods, though.